src/org/jetbrains/sbt/project/ExternalSourceRootResolution.scala

```diff
@@ -185,8 +185,12 @@ trait ExternalSourceRootResolution { self: SbtProjectResolver =>
     var counter = 1
 
     def nameFor(base: Option[File]) = {
-      val namedDirectory = if (base.exists(_.getName == "shared")) base.flatMap(_.parent) else base
-      val prefix = namedDirectory.map(_.getName + "-sources").getOrElse("shared-sources")
+      val namedDirectory = {
+        var d = base
+        while (d.exists(_.getName == "shared")) d = d.flatMap(_.parent)
```

Comment:
Ideally, it's better to avoid the "imperative" syntax and the mutable state (though local), for example, by defining `parents: Stream[File]` helper method in `RichFile`.